forwardproxy.go

```diff
@@ -695,8 +699,182 @@ type dialContexter interface {
 	DialContext(ctx context.Context, network, address string) (net.Conn, error)
 }
 
+func (h *Handler) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
+	if !d.Next() {
+		return d.ArgErr()
+	}
+	args := d.RemainingArgs()
+	if len(args) > 0 {
+		return d.ArgErr()
+	}
+	for nesting := d.Nesting(); d.NextBlock(nesting); {
+		subdirective := d.Val()
+		args := d.RemainingArgs()
+		switch subdirective {
+		case "basic_auth":
+			if len(args) != 2 {
+				return d.ArgErr()
+			}
+			if len(args[0]) == 0 {
+				return d.Err("empty usernames are not allowed")
+			}
+			// TODO: Evaluate policy of allowing empty passwords.
+			if strings.Contains(args[0], ":") {
+				return d.Err("character ':' in usernames is not allowed")
+			}
+			// TODO: Support multiple basicauths.
+			if h.BasicauthUser != "" || h.BasicauthPass != "" {
+				return d.Err("Multi-user basicauth is not supported")
+			}
+			h.BasicauthUser = args[0]
+			h.BasicauthPass = args[1]
+		case "ports":
+			if len(args) == 0 {
+				return d.ArgErr()
+			}
+			if len(h.WhitelistedPorts) != 0 {
+				return d.Err("ports subdirective specified twice")
+			}
+			h.WhitelistedPorts = make([]int, len(args))
+			for i, p := range args {
+				intPort, err := strconv.Atoi(p)
+				if intPort <= 0 || intPort > 65535 || err != nil {
+					return d.Err("ports are expected to be space-separated" +
+						" and in 0-65535 range. Got: " + p)
+				}
+				h.WhitelistedPorts[i] = intPort
+			}
+		case "hide_ip":
+			if len(args) != 0 {
+				return d.ArgErr()
+			}
+			h.HideIP = true
+		case "hide_via":
+			if len(args) != 0 {
+				return d.ArgErr()
+			}
+			h.HideVia = true
+		case "probe_resistance":
+			if len(args) > 1 {
+				return d.ArgErr()
+			}
+			if len(args) == 1 {
+				lowercaseArg := strings.ToLower(args[0])
+				if lowercaseArg != args[0] {
+					log.Println("WARNING: secret domain appears to have uppercase letters in it, which are not visitable")
```

Comment:
I just realized this file is using the standard logger still (`log` package).

In Caddy 2 we don't use the standard logger. Use a Logger which you can obtain from Provision: https://pkg.go.dev/github.com/caddyserver/caddy/v2#Context.Logger - would you have a moment to replace all `log.*` with the structured logger? I hadn't gotten around to that yet but it might be a good time to do that.

The `"log"` package import should disappear.